### PR TITLE
make warnings/logging capture threadsafe

### LINKED ISSUE #106: Logging capture runs afoul of thread safety
```
Exception in thread Thread-1991 (_log_start_if_long):
Traceback (most recent call last):
  File "/home/andreas/pack/mambaforge/envs/ceesd/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "/home/andreas/pack/mambaforge/envs/ceesd/lib/python3.11/threading.py", line 975, in run
    self._target(*self._args, **self._kwargs)
  File "/home/andreas/src/pytools/pytools/__init__.py", line 2617, in _log_start_if_long
    self.logger.log(
  File "/home/andreas/pack/mambaforge/envs/ceesd/lib/python3.11/logging/__init__.py", line 1559, in log
    self._log(level, msg, args, **kwargs)
  File "/home/andreas/pack/mambaforge/envs/ceesd/lib/python3.11/logging/__init__.py", line 1634, in _log
    self.handle(record)
  File "/home/andreas/pack/mambaforge/envs/ceesd/lib/python3.11/logging/__init__.py", line 1644, in handle
    self.callHandlers(record)
  File "/home/andreas/pack/mambaforge/envs/ceesd/lib/python3.11/logging/__init__.py", line 1706, in callHandlers
    hdlr.handle(record)
  File "/home/andreas/pack/mambaforge/envs/ceesd/lib/python3.11/logging/__init__.py", line 978, in handle
    self.emit(record)
  File "/home/andreas/src/logpyle/logpyle/__init__.py", line 662, in emit
    self.mgr.db_conn.execute(
sqlite3.ProgrammingError: SQLite objects created in a thread can only be used in that same thread. The object was created in thread id 139805741225792 and this is thread id 139798635280064.
```

### PR DESCRIPTION
**Pull request by matthiasdiener**
Fixes #106. Alternative to #107.

### CODE REVIEW

**Comment by matthiasdiener** [comment]
What do you think @inducer ?

File: logpyle/__init__.py [hunk 1]
```diff
@@ -919,7 +932,29 @@ def _commit(self) -> None:
...
+                "insert into logging values (?,?,?,?,?,?,?)",
+                (self.rank, log.tick_count, log.time,
+                log.category, log.message, log.filename,
+                log.lineno))
+
+        self.logging_data = []
+
+    def save_warnings(self) -> None:
+        for w in self.warning_data:
+            self.db_conn.execute(
```
Comment (inducer) [inline logpyle/__init__.py:947]: It seems support for old schema versions was lost here.

Comment (matthiasdiener) [reply logpyle/__init__.py:947]: I can restore that support if you prefer, but is it useful to be able to _write_ older schema version?

Comment (inducer) [reply logpyle/__init__.py:947]: I'm not opposed to entirely *dropping* that support even, but there have to be error messages that clarify what happened, not failures/incorrect execution of `INSERT` somewhere in the guts.

Comment (matthiasdiener) [reply logpyle/__init__.py:957]: What do you think of 30ed3fb49f9e707ac506f8d5009a55fcad780686?

Comment (matthiasdiener) [reply logpyle/__init__.py:957]: Ok, I added this as issue #109.

File: logpyle/__init__.py [hunk 2]
```diff
@@ -608,33 +618,44 @@ def __init__(self, filename: Optional[str] = None, mode: str = "r",
...
+                raise ValueError("Warnings capture needs at least schema_version 3, "
+                                f" got {self.schema_version}")
             self.capture_warnings(True)
 
+        self.logging_data: List[_LogWarningInfo] = []
         self.logging_handler: Optional[logging.Handler] = None
         if capture_logging:
+            if self.schema_version < 3:
+                raise ValueError("Logging capture needs at least schema_version 3, "
+                                f" got {self.schema_version}")
```
[the remark this answers is not in the document]
Comment (matthiasdiener) [reply logpyle/__init__.py]: I moved them in cd8301f

**Review by inducer — APPROVED** [review]

**Pull request merged**